File: web/PersonalOS.Web/src/app/core/study/study.models.ts
import { IsoLocalDate } from '../time/local-date';

/** Where a study project stands, exactly as the API spells it. */
export type StudyProjectStatus = 'active' | 'paused' | 'completed';

/** What kind of material a resource points to, exactly as the API spells it. */
export type StudyResourceType = 'audio' | 'video' | 'pdf' | 'exam' | 'article' | 'other';

/** Options offered by the project status picker. */
export const STUDY_PROJECT_STATUSES: readonly {
  value: StudyProjectStatus;
  label: string;
}[] = [
  { value: 'active', label: 'Active' },
  { value: 'paused', label: 'Paused' },
  { value: 'completed', label: 'Completed' },
];

/** Options offered by the resource type picker. */
export const STUDY_RESOURCE_TYPES: readonly {
  value: StudyResourceType;
  label: string;
}[] = [
  { value: 'audio', label: 'Audio' },
  { value: 'video', label: 'Video' },
  { value: 'pdf', label: 'PDF' },
  { value: 'exam', label: 'Exam' },
  { value: 'article', label: 'Article' },
  { value: 'other', label: 'Other' },
];

/**
 * A reference to study material.
 *
 * `externalUrl` is always an `http` or `https` address: the server rejects any other scheme
 * before storing it. Templates still render it only through `[href]`, never as HTML.
 */
export interface StudyResource {
  readonly id: string;
  readonly title: string;
  readonly resourceType: StudyResourceType;
  readonly externalUrl: string | null;
  readonly notes: string | null;
}

/** A subject or learning project with its material. */
export interface StudyProject {
  readonly id: string;
  readonly name: string;
  readonly description: string | null;
  readonly status: StudyProjectStatus;
  readonly resources: readonly StudyResource[];
}

/** One recorded block of studying. */
export interface StudySession {
  readonly id: string;
  readonly studyProjectId: string;
  readonly projectName: string;
  readonly localDate: IsoLocalDate;
  readonly startTime: string | null;
  readonly durationMinutes: number;
  readonly summary: string | null;
  readonly progressNote: string | null;
}

/** Values sent for one study resource. */
export interface StudyResourceRequest {
  readonly title: string;
  readonly resourceType: StudyResourceType;
  readonly externalUrl: string | null;
  readonly notes: string | null;
}

/** Values sent when creating or editing a study project. */
export interface SaveStudyProjectRequest {
  readonly name: string;
  readonly description: string | null;
  readonly status: StudyProjectStatus;
  readonly resources: readonly StudyResourceRequest[];
}

/** Values sent when recording or editing a study session. */
export interface SaveStudySessionRequest {
  readonly studyProjectId: string;
  readonly localDate: IsoLocalDate;
  readonly startTime: string | null;
  readonly durationMinutes: number;
  readonly summary: string | null;
  readonly progressNote: string | null;
}

File: src/PersonalOS.Application/Study/StudyModels.cs
using PersonalOS.Domain.Study;

namespace PersonalOS.Application.Study;

/// <summary>
/// A reference to study material.
/// </summary>
/// <param name="Id">Resource identifier.</param>
/// <param name="Title">How the user named the material.</param>
/// <param name="ResourceType">What kind of material it is.</param>
/// <param name="ExternalUrl">Optional <c>http</c> or <c>https</c> address.</param>
/// <param name="Notes">Optional note.</param>
public sealed record StudyResourceRecord(
    Guid Id,
    string Title,
    StudyResourceType ResourceType,
    string? ExternalUrl,
    string? Notes)
{
    /// <summary>
    /// Projects a domain entity onto the application record.
    /// </summary>
    /// <param name="resource">Domain entity.</param>
    public static StudyResourceRecord FromEntity(StudyResource resource)
    {
        ArgumentNullException.ThrowIfNull(resource);

        return new StudyResourceRecord(
            resource.Id,
            resource.Title,
            resource.ResourceType,
            resource.ExternalUrl,
            resource.Notes);
    }
}

/// <summary>
/// Values a client may supply for one study resource.
/// </summary>
/// <param name="Title">How the user named the material.</param>
/// <param name="ResourceType">What kind of material it is.</param>
/// <param name="ExternalUrl">Optional web address.</param>
/// <param name="Notes">Optional note.</param>
public sealed record StudyResourceInput(
    string? Title,
    StudyResourceType ResourceType,
    string? ExternalUrl,
    string? Notes);

/// <summary>
/// A subject or learning project with its material.
/// </summary>
/// <param name="Id">Project identifier.</param>
/// <param name="Name">Subject name.</param>
/// <param name="Description">Optional longer text.</param>
/// <param name="Status">Where the project stands.</param>
/// <param name="Resources">Material attached to the project.</param>
public sealed record StudyProjectRecord(
    Guid Id,
    string Name,
    string? Description,
    StudyProjectStatus Status,
    IReadOnlyList<StudyResourceRecord> Resources)
{
    /// <summary>
    /// Projects a domain entity onto the application record.
    /// </summary>
    /// <param name="project">Domain entity.</param>
    public static StudyProjectRecord FromEntity(StudyProject project)
    {
        ArgumentNullException.ThrowIfNull(project);

        return new StudyProjectRecord(
            project.Id,
            project.Name,
            project.Description,
            project.Status,
            [.. project.Resources.Select(StudyResourceRecord.FromEntity)]);
    }
}

/// <summary>
/// Values a client may supply when creating or editing a study project.
/// </summary>
/// <param name="Name">Subject name.</param>
/// <param name="Description">Optional longer text.</param>
/// <param name="Status">Where the project stands.</param>
/// <param name="Resources">Material attached to the project.</param>
public sealed record StudyProjectInput(
    string? Name,
    string? Description,
    StudyProjectStatus Status,
    IReadOnlyList<StudyResourceInput>? Resources);

/// <summary>
/// One recorded block of studying.
/// </summary>
/// <param name="Id">Session identifier.</param>
/// <param name="StudyProjectId">Project that was studied.</param>
/// <param name="ProjectName">Name of that project, so the client needs no second request.</param>
/// <param name="LocalDate">The owner's local calendar day.</param>
/// <param name="StartTime">Optional local start time.</param>
/// <param name="DurationMinutes">How long the session lasted.</param>
/// <param name="Summary">What was studied.</param>
/// <param name="ProgressNote">Where the user now stands.</param>
public sealed record StudySessionRecord(
    Guid Id,
    Guid StudyProjectId,
    string ProjectName,
    DateOnly LocalDate,
    TimeOnly? StartTime,
    int DurationMinutes,
    string? Summary,
    string? ProgressNote)
{
    /// <summary>
    /// Projects a domain entity onto the application record.
    /// </summary>
    /// <param name="session">Domain entity.</param>
    /// <param name="projectName">Name of the project the session belongs to.</param>
    public static StudySessionRecord FromEntity(StudySession session, string projectName)
    {
        ArgumentNullException.ThrowIfNull(session);

        return new StudySessionRecord(
            session.Id,
            session.StudyProjectId,
            projectName,
            session.LocalDate,
            session.StartTime,
            session.DurationMinutes,
            session.Summary,
            session.ProgressNote);
    }
}

/// <summary>
/// Values a client may supply when recording or editing a study session.
/// </summary>
/// <param name="StudyProjectId">Project that was studied.</param>
/// <param name="LocalDate">The owner's local calendar day.</param>
/// <param name="StartTime">Optional local start time.</param>
/// <param name="DurationMinutes">How long the session lasted.</param>
/// <param name="Summary">What was studied.</param>
/// <param name="ProgressNote">Where the user now stands.</param>
public sealed record StudySessionInput(
    Guid? StudyProjectId,
    DateOnly? LocalDate,
    TimeOnly? StartTime,
    int? DurationMinutes,
    string? Summary,
    string? ProgressNote);
